src/java/org/apache/cassandra/streaming/async/StreamingInboundHandler.java

```diff
@@ -0,0 +1,236 @@
+ * and write those out to new sstable files. Because deserialization is a blocking affair, we can't block
+ * the netty event loop. Thus we have a background thread perform all the blocking deserialization.
+ *
+ *
+ *  * // TODO:JEB document
+ * - netty low/high water marks & how it relates to our use of auto-read
+ * - netty auto-read, and why we do it on the event loop (keeps things simple from the netty dispatch POV)
+ */
+public class StreamingInboundHandler extends ChannelInboundHandlerAdapter
+{
+    private static final Logger logger = LoggerFactory.getLogger(StreamingInboundHandler.class);
+
+    public static final int CHECKSUM_LENGTH = Integer.BYTES;
+
+    private static final int AUTO_READ_LOW_WATER_MARK = 1 << 15; // 1 << 19 = 512Kb
+    private static final int AUTO_READ_HIGH_WATER_MARK = 1 << 16; // 1 << 22 = 4Mb
+
+    enum State { START, MESSAGE_PAYLOAD, FILE_TRANSFER_HEADER_PAYLOAD, FILE_TRANSFER_PAYLOAD, CLOSED }
+
+    private final InetSocketAddress remoteAddress;
+    private final int protocolVersion;
+
+    private State state;
+
+    /**
+     * A collection of {@link ByteBuf}s that are yet to be processed. Incoming buffers are first dropped into this
+     * structure, and then consumed.
+     */
+    private ByteBufReadableByteChannel bufChannel;
+
+    /**
+     * A background thread that performs the deserialization of the sstable data.
+     */
+    private Thread blockingIOThread;
+
+    public StreamingInboundHandler(InetSocketAddress remoteAddress, int protocolVersion)
+    {
+        this.remoteAddress = remoteAddress;
+        this.protocolVersion = protocolVersion;
+        state = State.START;
+    }
+
+    @Override
+    @SuppressWarnings("resource")
+    public void handlerAdded(ChannelHandlerContext ctx)
+    {
+        bufChannel = new ByteBufReadableByteChannel(AUTO_READ_LOW_WATER_MARK, AUTO_READ_HIGH_WATER_MARK, ctx.channel().config());
+        blockingIOThread = new FastThreadLocalThread(new DeserializingSstableTask(ctx),
+                                                     String.format("Stream-Inbound--%s-%s", ctx.channel().id(), remoteAddress.toString()));
+        blockingIOThread.setDaemon(true);
+        blockingIOThread.start();
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object message) throws Exception
+    {
+        if (state == State.CLOSED)
+        {
+            ReferenceCountUtil.release(message);
+            return;
+        }
+
+        if (!(message instanceof ByteBuf))
+        {
+            ctx.fireChannelRead(message);
+            return;
+        }
+
+        bufChannel.append((ByteBuf) message);
+    }
+
+    @Override
+    public void channelInactive(ChannelHandlerContext ctx)
+    {
+        close();
+        ctx.fireChannelInactive();
+    }
+
+    private void close()
+    {
+        if (state == State.CLOSED)
+            return;
+
+        state = State.CLOSED;
+        blockingIOThread.interrupt();
```

Comment:
Never interrupt! If it's blocked on something close the thing it's blocked on or signal and that can wake it up.

It could throw interrupted exception anywhere and there is no guarantee it's where you expect or that it is handled correctly.